### PR TITLE
Add binary search MASM example

### PR DESCRIPTION
**Pull request by polydez**
This is a Miden Assembly example code, which performs binary search for a given value in a given ascending array.  All input data should be provided in the `operands_stack` of intputs-file. If the `operands_stack` is empty, tests will run.

### CODE REVIEW

File: examples/bsearch.masm [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+#! For input and output see documentation for `proc.binary_search_stack`.
+#!
+#! If input values are omitted or `count = 0`, tests will run.
+
+const.STARTING_MEMORY_ADDRESS=0
+
+#! Binary searches for value `val` in the memory addresses range [start..end), sorted in ascending order.
+#!
+#! Input: [start, end, val, ...]
+#! Output: [found (true/false), addr] where `found` is searching result (`true`/`false`),
```
Comment (Dominik1999) [inline examples/bsearch.masm:13]: > where `found` is searching result (`true`/`false`)

nit: maybe better "where `found` indicates the result

File: examples/bsearch.masm [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+#!         `addr` is an address of value (if it was found), or an address where value should be inserted.
```
Comment (Dominik1999) [inline examples/bsearch.masm:14]: > `addr` is an address of value (if it was found), or an address where value should be inserted.

nit: can you add address in memory?

File: examples/bsearch.masm [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+#! Input: [count, arr[0], arr[1], ..., arr[count - 1], val],
+#!        where:
+#!            `count` is number of elements in the array,
+#!            `arr[0]`, `arr[1]`, ..., `arr[count - 1]` are array elements,
+#!            `val` is value to search.
+#! Output: [found, index] where `found` is searching result (`true`/`false`),
+#!         index `[0, count)` of `val`, if `val` was found; otherwise index in array, where `val` should be inserted.
+proc.binary_search_stack
+    push.STARTING_MEMORY_ADDRESS            # [STARTING_MEMORY_ADDRESS, count, arr[0], arr[1], ..., arr[count - 1], val, ...]
+    exec.read_stack_to_memory               # [end, val, ...]
```
Comment (Dominik1999) [inline examples/bsearch.masm:113]: should it be `write_stack_to_memort`?

File: examples/bsearch.masm [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    drop                                    # [addr, found, val, ...]
+    swap.2                                  # [val, found, addr, ...]
+    drop                                    # [found, addr, ...]
+end
+
+#! Reads `count` values (`arr[0], arr[1], ..., arr[count - 1]`) from the stack into the memory starting from address `addr`.
+#!
+#! Input: [addr, count, arr[0], arr[1], ..., arr[count - 1], ...]
+#! Output: [end, ...] where `end` is addr next vacant memory address, thus array occupies addresses `[addr..end)` in the memory
+proc.read_stack_to_memory
```
Comment (Dominik1999) [inline examples/bsearch.masm:83]: same here, should it be `write_stack_to_memory`?

File: examples/bsearch.masm [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    u32assert2                              # [addr, count, arr[0], arr[1], ...]
+    dup.1                                   # [count, addr, c, arr[0], arr[1], ...]
```
Comment (Dominik1999) [inline examples/bsearch.masm:85]: nit: `[count, addr, count, arr[0], arr[1], ...]`

File: examples/bsearch.masm (hunk 4 above)
Comment (Dominik1999) [inline examples/bsearch.masm:83]: Nice procedure to store arbitrary length

File: examples/bsearch.masm [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+proc.binary_search
+    # Check if `start < end` then start search, otherwise prepare "not found" result and exit
+    u32assert2                              # [start, end, val, ...]
+    dup                                     # [start, start, end, val, ...]
+    dup.2                                   # [end, start, start, end, val, ...]
+    u32checked_lt                           # [start < end, start, end, val, ...]
+    if.true                                 # [start, end, val, ...]
+        push.1                              # [1, start, end, val, ...]
+    else                                    # [start, end, val, ...]
+        # Push "not found" flag and flag to skip the loop
```
Comment (Dominik1999) [inline examples/bsearch.masm:24]: This would be an error, right? 

You could use `assert.err=123` and break.

File: examples/bsearch.masm [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        push.0.0                            # [0, 0, addr, end, val, ...], addr = start
+    end
+    while.true                              # [start, end, val, ...]
+        # Calculate middle address `middle`
+        dup.1                               # [end, start, end, val, ...]
+        dup.1                               # [start, end, start, end, val, ...]
+        u32wrapping_sub                     # [count, start, end, val, ...], count = end - start
+        u32checked_div.2                    # [count / 2, start, end, val, ...]
```
Comment (Dominik1999) [inline examples/bsearch.masm:32]: nit: can you add in the comment what happens if `count` is odd?

File: examples/bsearch.masm [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+        dup                                 # [memory[middle], memory[middle], middle, start, end, val, ...]
+        dup.5                               # [val, memory[middle], memory[middle], middle, start, end, val, ...]
+        eq                                  # [memory[middle] == val, memory[middle], middle, start, end, val, ...]
+        if.true                             # [memory[middle], middle, start, end, val, ...]
+            drop                            # [middle, start, end, val, ...]
+            swap                            # [start, middle, end, val, ...]
+            drop                            # [middle, end, val, ...]
+            # Prepare "found" flag and exit the loop
+            push.1.0                        # [0, 1, addr, end, val, ...], addr = middle
```
Comment (Dominik1999) [inline examples/bsearch.masm:47]: I like the usage of the flags like this

File: examples/bsearch.masm [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    call.test_gaps_left_found
+    call.test_gaps_right_not_found
+    call.test_gaps_left_not_found
+end
+
+begin
+    dup                                     # [count, count, ...]
+    neq.0                                   # [count != 0, count, ...]
+    if.true                                 # [count, arr[0], arr[1], ..., arr[count - 1], val, ...]
+        exec.binary_search_stack            # [found, index]
```
Comment (Dominik1999) [inline examples/bsearch.masm:203]: can you maybe get rid of the `found` in the final output?

You could return all `0` if you didn't find the value and the pos of the stack if you found it

File: examples/bsearch.masm [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    push.STARTING_MEMORY_ADDRESS            # [start, end, val, ...], start = STARTING_MEMORY_ADDRESS
+    exec.binary_search                      # [found, addr, ...]
+
+    swap                                    # [addr, found, ...]
+    push.STARTING_MEMORY_ADDRESS            # [STARTING_MEMORY_ADDRESS, addr, found, ...]
+    u32wrapping_sub                         # [index, found, ...], index = addr - STARTING_MEMORY_ADDRESS
+    swap                                    # [found, index, ...]
+end
+
+proc.test_empty
```
Comment (Dominik1999) [inline examples/bsearch.masm:124]: This is super nice that you made tests!

**Review by Dominik1999 — COMMENTED** [review]
Thanks Denis! This is a super cool MASM program. I left a couple of comments in line. 

One point: We try to use the convention now that we first write what the next line is about and then indicate the final stack after the operations / line with `# => [...]`

like here 

```
    # compute the pointer at which we should stop iterating
    dup.1 add
    # => [end_ptr, ptr, ...]
```

Can you change that quickly?
(I know it's not consistent yet with all the examples. I will change that.

We should combine this with the Bubble Sort of @phklive when merged.

File: examples/bsearch.masm (hunk 6 above)
Comment (polydez) [reply examples/bsearch.masm:24]: This is not an error. Binary search is used not only for searching, but also for adding new values into a sorted array. In a case when value wasn't found, the algorithm must return an index where this value should be inserted.

File: examples/bsearch.masm (hunk 9 above)
Comment (polydez) [reply examples/bsearch.masm:203]: I think, it wouldn't be a good idea to return indexes of the stack. We work with index in array, which is passed through the stack. If value is not found, `index == 0` would conflict with the index of 0th element. I tried to use special index like `0xFFFFFFFF` for such cases, but it made the code worse.

Comment (Dominik1999) [reply examples/bsearch.masm]: agreed, thanks for the explanation

File: examples/bsearch.inputs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+{
+    "note": [
```
Comment (Dominik1999) [inline examples/bsearch.inputs:2]: Can you move those "note" comments to the beginning of the bsearch.masm file? It doesn't work on the playground (yet) having "notes" in the input

**Review by Dominik1999 — APPROVED** [review]
Looks good to me, thanks!

**Comment by phklive** [comment]
Congrats!

**Pull request merged**